File: backend/src/modules/bale_monitor/baleClient.js
const axios = require('axios');

class BaleClient {
  constructor(token, timeoutMs = 10000, retries = 2) {
    this.baseUrl = `https://tapi.bale.ai/bot${token}`;
    this.timeoutMs = timeoutMs;
    this.retries = Math.max(1, retries);
  }

  async sendMessage(chatId, text) {
    return this._call('sendMessage', { chat_id: chatId, text });
  }

  async setWebhook(url) {
    return this._call('setWebhook', { url });
  }

  async deleteWebhook() {
    return this._call('deleteWebhook', {});
  }

  async getWebhookInfo() {
    return this._call('getWebhookInfo', {});
  }

  async getUpdates(offset, timeout = 0) {
    return this._call('getUpdates', { offset, timeout });
  }

  async _call(method, payload) {
    const url = `${this.baseUrl}/${method}`;

    let lastErr;
    for (let attempt = 1; attempt <= this.retries; attempt += 1) {
      try {
        const { data } = await axios.post(url, payload, {
          timeout: this.timeoutMs,
          headers: { 'Content-Type': 'application/json' }
        });
        if (data && data.ok === false) {
          throw new Error(`Bale API error: ${JSON.stringify(data)}`);
        }
        return data;
      } catch (err) {
        lastErr = err;
        if (attempt < this.retries) {
          await new Promise((resolve) => setTimeout(resolve, 250));
        }
      }
    }
    throw lastErr;
  }
}

module.exports = { BaleClient };

File: uptime/modules/bale_monitor/bale_client.py
from __future__ import annotations

import logging
import time

import requests

logger = logging.getLogger(__name__)


class BaleClient:
    def __init__(self, token: str, timeout_seconds: int = 10, retry_attempts: int = 2) -> None:
        self.base_url = f"https://tapi.bale.ai/bot{token}"
        self.timeout_seconds = timeout_seconds
        self.retry_attempts = max(1, retry_attempts)

    def send_message(self, chat_id: int, text: str) -> dict:
        url = f"{self.base_url}/sendMessage"
        payload = {"chat_id": chat_id, "text": text}
        return self._post_with_retry(url, payload)

    def _post_with_retry(self, url: str, payload: dict) -> dict:
        last_error: Exception | None = None
        for attempt in range(1, self.retry_attempts + 1):
            try:
                response = requests.post(url, json=payload, timeout=self.timeout_seconds)
                response.raise_for_status()
                data = response.json()
                if data.get("ok") is False:
                    raise RuntimeError(f"Bale API error: {data}")
                return data
            except Exception as exc:  # noqa: BLE001
                last_error = exc
                logger.warning("Bale API attempt %s/%s failed: %s", attempt, self.retry_attempts, exc)
                if attempt < self.retry_attempts:
                    time.sleep(0.25)
        raise RuntimeError(f"Bale API request failed after retries: {last_error}")
